File: python/tests/test_m98_assembly_graph.py
"""M98 AssemblyGraph acceptance tests.

GFA parse/emit byte-exactness, the ``/study/assembly_graphs`` storage
round-trip, the ``opt_assembly_graph`` feature flag, and the
sequences-channel codec selection.

Mirrors:
    objc/Tests/TestM98AssemblyGraph.m
    java/src/test/java/.../M98AssemblyGraphTest.java
"""
from __future__ import annotations

import os
from pathlib import Path

import pytest

# The synthetic full-surface GFA the Phase 0 proof used: every GFA 1.x
# line type, sequence-less S records, tag stacks, a comment, a
# hifiasm-style A extension, interleaved ordering. Kept in lockstep
# with the ObjC and Java fixtures.
_SYNTH_LINES = [
    "H\tVN:Z:1.0",
    "# produced by the m98 synthetic generator",
    "S\tutg000001l\tACGTACGTACGTNNNACGT\tLN:i:19\trd:i:12",
    "A\tutg000001l\t0\t+\tread_00001\t0\t19\tid:i:0\tHG:A:a",
    "S\tutg000002l\t*\tLN:i:5000",
    "L\tutg000001l\t+\tutg000002l\t-\t15M\tL1:i:4985",
    "S\tutg000003c\tGGGGCCCCTTTTAAAA\tLN:i:16",
    "L\tutg000002l\t-\tutg000003c\t+\t*",
    "C\tutg000001l\t+\tutg000003c\t-\t2\t14M\tNM:i:0",
    "P\tscaffold_1\tutg000001l+,utg000002l-,utg000003c+\t15M,*\tXX:Z:demo",
    "L\tutg000003c\t+\tutg000001l\t+\t0M",
]


def _synth_gfa() -> bytes:
    return ("\n".join(_SYNTH_LINES) + "\n").encode()


# -------------------------------------------- per-AU encryption + signatures


def _write_graph_tio(tmp_path: Path, name: str = "m98e.tio") -> tuple:
    from ttio.importers.gfa import GfaReader
    from ttio.spectral_dataset import SpectralDataset

    src = _synth_gfa()
    g = GfaReader.graph_from_bytes(src)
    p = tmp_path / name
    SpectralDataset.write_minimal(
        p, title="M98", isa_investigation_id="ISA-M98", runs={},
        assembly_graphs={"graph_0001": g},
    )
    return p, src


def test_per_au_encrypt_decrypt_in_place(tmp_path: Path):
    """Workplan acceptance: a per-AU-encrypted graph decrypts in
    place. Mechanism check: after encrypt the plaintext sequences
    channel is GONE and the segments compound exists -- a no-op
    walker would still pass a bare round-trip."""
    import h5py

    from ttio.encryption_per_au import (
        decrypt_per_au_in_place,
        encrypt_per_au,
    )
    from ttio.spectral_dataset import SpectralDataset

    p, src = _write_graph_tio(tmp_path)
    key = bytes(range(32))
    encrypt_per_au(str(p), key)

    with h5py.File(p, "r") as f:
        seg = f["study/assembly_graphs/graph_0001/segments"]
        assert "sequences" not in seg, "plaintext channel not removed"
        assert "sequences_segments" in seg, "segments compound missing"
        alg = seg.attrs["sequences_algorithm"]
        assert alg in ("aes-256-gcm", b"aes-256-gcm")

    # The wrong key must not decrypt.
    with pytest.raises(Exception):
        decrypt_per_au_in_place(str(p), bytes(32))

    decrypt_per_au_in_place(str(p), key)
    with h5py.File(p, "r") as f:
        seg = f["study/assembly_graphs/graph_0001/segments"]
        assert "sequences" in seg and "sequences_segments" not in seg

    ds = SpectralDataset.open(p)
    try:
        assert "opt_per_au_encryption" not in ds.feature_flags.features
        assert ds.assembly_graphs["graph_0001"].gfa_bytes() == src
    finally:
        ds.close()


def test_signatures_cover_sequences_channel(tmp_path: Path):
    """v2: signatures cover segments/sequences the way they cover
    genomic-run channels; a tampered channel fails verification."""
    import h5py

    from ttio.signatures import (
        sign_assembly_graph,
        verify_assembly_graph,
    )

    p, _src = _write_graph_tio(tmp_path, "m98s.tio")
    key = b"k" * 32
    with h5py.File(p, "r+") as f:
        gg = f["study/assembly_graphs/graph_0001"]
        sigs = sign_assembly_graph(gg, key)
        assert "segments/sequences" in sigs
        assert sigs["segments/sequences"].startswith("v2:")
        assert verify_assembly_graph(gg, key)
        assert not verify_assembly_graph(gg, b"x" * 32)

    with h5py.File(p, "r+") as f:
        seq = f["study/assembly_graphs/graph_0001/segments/sequences"]
        first = int(seq[0])
        seq[0] = (first + 1) % 256
    with h5py.File(p, "r") as f:
        gg = f["study/assembly_graphs/graph_0001"]
        assert not verify_assembly_graph(gg, key), (
            "tampered sequences channel must fail verification")


# ---------------------------------------------------------------- parse/emit


def test_gfa_parse_emit_byte_exact():
    from ttio.exporters.gfa import GfaWriter
    from ttio.importers.gfa import GfaReader

    src = _synth_gfa()
    g = GfaReader.graph_from_bytes(src)
    assert (len(g.segments), len(g.links), len(g.paths), len(g.extras)) \
        == (3, 3, 1, 4)
    assert g.gfa_version == "1.0"
    assert g.segments[1].sequence is None  # '*' parses as None
    assert GfaWriter.data_for_graph(g) == src

    # The no-final-newline variant round-trips too.
    no_nl = src[:-1]
    g2 = GfaReader.graph_from_bytes(no_nl)
    assert g2.final_newline is False
    assert GfaWriter.data_for_graph(g2) == no_nl


# ------------------------------------- storage round-trip (memory provider)


def test_storage_round_trip_memory_provider():
    from ttio.assembly import AssemblyGraph, write_assembly_graph
    from ttio.importers.gfa import GfaReader
    from ttio.providers import open_provider

    src = _synth_gfa()
    g = GfaReader.graph_from_bytes(src)
    sp = open_provider(
        f"memory://m98-{os.getpid()}", provider="memory", mode="w")
    try:
        study = sp.root_group().create_group("study")
        write_assembly_graph(g, "g0", study)

        # Duplicate names are rejected.
        with pytest.raises(ValueError, match="already exists"):
            write_assembly_graph(g, "g0", study)

        gg = study.open_group("assembly_graphs").open_group("g0")
        opened = AssemblyGraph.open(gg, "g0")
        assert opened.gfa_bytes() == src
    finally:
        sp.close()


# ------------------------------------ write_minimal + reopen + feature flag


def test_write_minimal_flag_and_accessor(tmp_path: Path):
    from ttio.importers.gfa import GfaReader
    from ttio.spectral_dataset import SpectralDataset

    src = _synth_gfa()
    g = GfaReader.graph_from_bytes(src)
    p = tmp_path / "m98.tio"
    SpectralDataset.write_minimal(
        p, title="M98", isa_investigation_id="ISA-M98", runs={},
        assembly_graphs={"graph_0001": g},
    )

    ds = SpectralDataset.open(p)
    try:
        assert "opt_assembly_graph" in ds.feature_flags.features
        opened = ds.assembly_graphs["graph_0001"]
        assert opened.gfa_version == "1.0"
        assert opened.final_newline is True
        assert opened.gfa_bytes() == src
    finally:
        ds.close()


def test_write_minimal_graphless_has_no_flag(tmp_path: Path):
    # A file without graphs has neither the flag nor the subtree.
    from ttio.spectral_dataset import SpectralDataset

    p = tmp_path / "plain.tio"
    SpectralDataset.write_minimal(
        p, title="M98", isa_investigation_id="ISA-M98", runs={})
    ds = SpectralDataset.open(p)
    try:
        assert "opt_assembly_graph" not in ds.feature_flags.features
        assert ds.assembly_graphs == {}
    finally:
        ds.close()


# ---------------------------------------- sequences-channel codec selection


def test_sequences_channel_base_pack_engaged():
    """Mechanism check: the ACGT channel is BASE_PACK-encoded in the
    store (``@compression`` = 6 and stored < raw), not merely
    round-tripped -- a raw pass-through would satisfy a byte-compare."""
    from ttio.assembly import AssemblyGraph, write_assembly_graph
    from ttio.importers.gfa import GfaReader
    from ttio.providers import open_provider

    bases = "ACGT" * 2048  # 8,192 bases
    src = (f"S\tu1\t{bases}\nL\tu1\t+\tu1\t-\t0M\n").encode()
    g = GfaReader.graph_from_bytes(src)
    sp = open_provider(
        f"memory://m98c-{os.getpid()}", provider="memory", mode="w")
    try:
        study = sp.root_group().create_group("study")
        write_assembly_graph(g, "g0", study)

        seq_ds = (
            study.open_group("assembly_graphs").open_group("g0")
            .open_group("segments").open_dataset("sequences"))
        assert int(seq_ds.get_attribute("compression")) == 6  # BASE_PACK
        stored = seq_ds.read()
        assert 0 < len(stored) < 8192

        opened = AssemblyGraph.open(
            study.open_group("assembly_graphs").open_group("g0"), "g0")
        assert opened.gfa_bytes() == src
    finally:
        sp.close()

File: objc/Tests/TestM98AssemblyGraph.m
// TestM98AssemblyGraph.m — M98: GFA parse/emit byte-exactness, the
// /study/assembly_graphs storage round-trip, the opt_assembly_graph
// feature flag, and the sequences-channel codec selection.
//
// Mirrors:
//   python/tests/test_m98_assembly_graph.py
//   java/src/test/java/.../M98AssemblyGraphTest.java
//
// SPDX-License-Identifier: LGPL-3.0-or-later

#import <Foundation/Foundation.h>
#import "Testing.h"
#import "Dataset/TTIOSpectralDataset.h"
#import "Dataset/TTIOSpectralDataset+AssemblyWrite.h"
#import "Assembly/TTIOWrittenAssemblyGraph.h"
#import "Assembly/TTIOGraphSegment.h"
#import "Assembly/TTIOGraphLink.h"
#import "Assembly/TTIOGraphPath.h"
#import "Assembly/TTIOAssemblyGraph.h"
#import "Import/TTIOGfaReader.h"
#import "Export/TTIOGfaWriter.h"
#import "HDF5/TTIOHDF5File.h"
#import "HDF5/TTIOHDF5Group.h"
#import "HDF5/TTIOFeatureFlags.h"
#import "Providers/TTIOMemoryProvider.h"
#import "Providers/TTIOProviderRegistry.h"
#import "Providers/TTIOStorageProtocols.h"
#import "Protection/TTIOPerAUFile.h"
#import "Protection/TTIOSignatureManager.h"
#include <unistd.h>

/** The synthetic full-surface GFA the Phase 0 proof used: every GFA
 *  1.x line type, sequence-less S records, tag stacks, a comment, a
 *  hifiasm-style A extension, interleaved ordering. Kept in lockstep
 *  with the Python and Java fixtures. */
static NSString *m98SynthGfa(void)
{
    NSArray *lines = @[
        @"H\tVN:Z:1.0",
        @"# produced by the m98 synthetic generator",
        @"S\tutg000001l\tACGTACGTACGTNNNACGT\tLN:i:19\trd:i:12",
        @"A\tutg000001l\t0\t+\tread_00001\t0\t19\tid:i:0\tHG:A:a",
        @"S\tutg000002l\t*\tLN:i:5000",
        @"L\tutg000001l\t+\tutg000002l\t-\t15M\tL1:i:4985",
        @"S\tutg000003c\tGGGGCCCCTTTTAAAA\tLN:i:16",
        @"L\tutg000002l\t-\tutg000003c\t+\t*",
        @"C\tutg000001l\t+\tutg000003c\t-\t2\t14M\tNM:i:0",
        @"P\tscaffold_1\tutg000001l+,utg000002l-,utg000003c+\t15M,*\tXX:Z:demo",
        @"L\tutg000003c\t+\tutg000001l\t+\t0M",
    ];
    return [[lines componentsJoinedByString:@"\n"]
        stringByAppendingString:@"\n"];
}

static NSString *m98TmpPath(NSString *tag)
{
    return [NSString stringWithFormat:@"/tmp/ttio_m98_%d_%@.tio",
            (int)getpid(), tag];
}

static void m98Rm(NSString *p)
{
    [[NSFileManager defaultManager] removeItemAtPath:p error:NULL];
}

// ── parse + emit byte-exactness ───────────────────────────────────

static void m98ParseEmit(void)
{
    NSData *src = [m98SynthGfa() dataUsingEncoding:NSUTF8StringEncoding];
    NSError *err = nil;
    TTIOWrittenAssemblyGraph *g = [TTIOGfaReader graphFromData:src
                                                         error:&err];
    PASS(g != nil, "M98 gfa: parse succeeds (%s)",
         [[err localizedDescription] UTF8String] ?: "");
    if (!g) return;
    PASS(g.segments.count == 3 && g.links.count == 3
             && g.paths.count == 1 && g.extras.count == 4,
         "M98 gfa: tables S=3 L=3 P=1 X=4 (got %lu/%lu/%lu/%lu)",
         (unsigned long)g.segments.count, (unsigned long)g.links.count,
         (unsigned long)g.paths.count, (unsigned long)g.extras.count);
    PASS([g.gfaVersion isEqualToString:@"1.0"],
         "M98 gfa: version from the H line");
    PASS(g.segments[1].sequence == nil,
         "M98 gfa: * sequence parses as nil");
    PASS([[TTIOGfaWriter dataForGraph:g] isEqualToData:src],
         "M98 gfa: emit(parse(x)) == x");

    // The no-final-newline variant round-trips too.
    NSData *noNl = [src subdataWithRange:NSMakeRange(0, src.length - 1)];
    TTIOWrittenAssemblyGraph *g2 = [TTIOGfaReader graphFromData:noNl
                                                          error:&err];
    PASS(g2 != nil && !g2.finalNewline
             && [[TTIOGfaWriter dataForGraph:g2] isEqualToData:noNl],
         "M98 gfa: no-final-newline round-trip");
}

// ── storage round-trip (memory provider) ──────────────────────────

static void m98StorageRoundTrip(void)
{
    NSData *src = [m98SynthGfa() dataUsingEncoding:NSUTF8StringEncoding];
    NSError *err = nil;
    TTIOWrittenAssemblyGraph *g = [TTIOGfaReader graphFromData:src
                                                         error:&err];
    if (!g) { PASS(NO, "M98 storage: parse failed"); return; }

    NSString *url = [NSString stringWithFormat:@"memory://m98-%d",
                     (int)getpid()];
    [TTIOMemoryProvider discardStore:url];
    id<TTIOStorageProvider> mem = [[TTIOProviderRegistry sharedRegistry]
        openURL:url mode:TTIOStorageOpenModeCreate provider:@"memory"
          error:&err];
    id<TTIOStorageGroup> study =
        [[mem rootGroupWithError:&err] createGroupNamed:@"study"
                                                  error:&err];
    PASS([TTIOSpectralDataset writeAssemblyGraph:g
                                           named:@"g0"
                                    toStudyGroup:study
                                           error:&err],
         "M98 storage: write to memory provider (%s)",
         [[err localizedDescription] UTF8String] ?: "");

    // Duplicate names are rejected.
    PASS(![TTIOSpectralDataset writeAssemblyGraph:g
                                            named:@"g0"
                                     toStudyGroup:study
                                            error:&err],
         "M98 storage: duplicate name rejected");

    id<TTIOStorageGroup> ag = [study openGroupNamed:@"assembly_graphs"
                                              error:&err];
    id<TTIOStorageGroup> gg = [ag openGroupNamed:@"g0" error:&err];
    TTIOAssemblyGraph *opened = [TTIOAssemblyGraph openFromGroup:gg
                                                            name:@"g0"
                                                           error:&err];
    PASS(opened != nil, "M98 storage: open succeeds (%s)",
         [[err localizedDescription] UTF8String] ?: "");
    if (!opened) return;
    NSData *emitted = [opened gfaDataWithError:&err];
    PASS([emitted isEqualToData:src],
         "M98 storage: memory-provider re-emission byte-exact");
}

// ── writeMinimal + reopen + feature flag ──────────────────────────

static void m98WriteMinimal(void)
{
    NSData *src = [m98SynthGfa() dataUsingEncoding:NSUTF8StringEncoding];
    NSError *err = nil;
    TTIOWrittenAssemblyGraph *g = [TTIOGfaReader graphFromData:src
                                                         error:&err];
    if (!g) { PASS(NO, "M98 writeMinimal: parse failed"); return; }

    NSString *path = m98TmpPath(@"wm");
    m98Rm(path);
    PASS([TTIOSpectralDataset writeMinimalToPath:path
                                           title:@"M98"
                             isaInvestigationId:@"ISA-M98"
                                         msRuns:@{}
                                     genomicRuns:nil
                                  assemblyGraphs:@{@"graph_0001": g}
                                 identifications:nil
                                 quantifications:nil
                               provenanceRecords:nil
                                           error:&err],
         "M98 writeMinimal: write succeeds (%s)",
         [[err localizedDescription] UTF8String] ?: "");

    // Feature flag present on the root.
    TTIOHDF5File *f = [TTIOHDF5File openReadOnlyAtPath:path error:&err];
    PASS(f != nil, "M98 writeMinimal: reopen HDF5");
    if (f) {
        PASS([TTIOFeatureFlags root:[f rootGroup]
                    supportsFeature:@"opt_assembly_graph"],
             "M98 writeMinimal: opt_assembly_graph flag set");
        [f close];
    }

    TTIOSpectralDataset *ds =
        [TTIOSpectralDataset readFromFilePath:path error:&err];
    PASS(ds != nil, "M98 writeMinimal: dataset reopens (%s)",
         [[err localizedDescription] UTF8String] ?: "");
    if (ds) {
        TTIOAssemblyGraph *opened = ds.assemblyGraphs[@"graph_0001"];
        PASS(opened != nil, "M98 writeMinimal: accessor finds the graph");
        if (opened) {
            PASS([opened.gfaVersion isEqualToString:@"1.0"]
                     && opened.finalNewline,
                 "M98 writeMinimal: attributes round-trip");
            NSData *emitted = [opened gfaDataWithError:&err];
            PASS([emitted isEqualToData:src],
                 "M98 writeMinimal: HDF5 re-emission byte-exact (%s)",
                 [[err localizedDescription] UTF8String] ?: "");
        }
        [ds closeFile];
    }

    // A file without graphs has neither the flag nor the subtree.
    NSString *plain = m98TmpPath(@"plain");
    m98Rm(plain);
    PASS([TTIOSpectralDataset writeMinimalToPath:plain
                                           title:@"M98"
                             isaInvestigationId:@"ISA-M98"
                                         msRuns:@{}
                                 identifications:nil
                                 quantifications:nil
                               provenanceRecords:nil
                                           error:&err],
         "M98 writeMinimal: graph-less write succeeds");
    TTIOHDF5File *pf = [TTIOHDF5File openReadOnlyAtPath:plain error:&err];
    if (pf) {
        PASS(![TTIOFeatureFlags root:[pf rootGroup]
                     supportsFeature:@"opt_assembly_graph"],
             "M98 writeMinimal: flag absent without graphs");
        [pf close];
    }
    TTIOSpectralDataset *pds =
        [TTIOSpectralDataset readFromFilePath:plain error:&err];
    PASS(pds != nil && pds.assemblyGraphs.count == 0,
         "M98 writeMinimal: graph-less file reads back empty dict");
    if (pds) [pds closeFile];
    m98Rm(path);
    m98Rm(plain);
}

// ── sequences-channel codec selection ─────────────────────────────

static void m98SequencesCodec(void)
{
    // ACGTN alphabet: the stored channel is BASE_PACK-encoded, so the
    // dataset must be SMALLER than the raw bases and carry
    // @compression. Mechanism check: parse, write, inspect the
    // dataset, reopen, byte-compare.
    NSMutableArray *lines = [NSMutableArray array];
    NSMutableString *bases = [NSMutableString string];
    for (int i = 0; i < 2048; i++) {
        [bases appendString:@"ACGT"];
    }
    [lines addObject:[NSString stringWithFormat:@"S\tu1\t%@", bases]];
    [lines addObject:@"L\tu1\t+\tu1\t-\t0M"];
    NSString *text = [[lines componentsJoinedByString:@"\n"]
        stringByAppendingString:@"\n"];
    NSData *src = [text dataUsingEncoding:NSUTF8StringEncoding];

    NSError *err = nil;
    TTIOWrittenAssemblyGraph *g = [TTIOGfaReader graphFromData:src
                                                         error:&err];
    NSString *url = [NSString stringWithFormat:@"memory://m98c-%d",
                     (int)getpid()];
    [TTIOMemoryProvider discardStore:url];
    id<TTIOStorageProvider> mem = [[TTIOProviderRegistry sharedRegistry]
        openURL:url mode:TTIOStorageOpenModeCreate provider:@"memory"
          error:&err];
    id<TTIOStorageGroup> study =
        [[mem rootGroupWithError:&err] createGroupNamed:@"study"
                                                  error:&err];
    PASS([TTIOSpectralDataset writeAssemblyGraph:g
                                           named:@"g0"
                                    toStudyGroup:study
                                           error:&err],
         "M98 codec: write succeeds");

    id<TTIOStorageGroup> segG = [[[study openGroupNamed:@"assembly_graphs"
                                                  error:NULL]
        openGroupNamed:@"g0" error:NULL]
        openGroupNamed:@"segments" error:NULL];
    id<TTIOStorageDataset> seqDs = [segG openDatasetNamed:@"sequences"
                                                    error:NULL];
    NSData *stored = (NSData *)[seqDs readAll:NULL];
    id codecAttr = [seqDs attributeValueForName:@"compression" error:NULL];
    PASS([codecAttr isKindOfClass:[NSNumber class]]
             && [codecAttr unsignedIntegerValue] == 6,
         "M98 codec: ACGT channel stored as BASE_PACK (@compression=6)");
    PASS(stored.length > 0 && stored.length < 8192,
         "M98 codec: 8,192 bases pack below 8,192 stored bytes (got %lu)",
         (unsigned long)stored.length);

    TTIOAssemblyGraph *opened = [TTIOAssemblyGraph
        openFromGroup:[[study openGroupNamed:@"assembly_graphs" error:NULL]
                          openGroupNamed:@"g0" error:NULL]
                 name:@"g0"
                error:&err];
    NSData *emitted = [opened gfaDataWithError:&err];
    if (![emitted isEqualToData:src]) {
        const uint8_t *a = src.bytes, *b = emitted.bytes;
        NSUInteger n = MIN(src.length, emitted.length), diff = n;
        for (NSUInteger i = 0; i < n; i++) {
            if (a[i] != b[i]) { diff = i; break; }
        }
        fprintf(stderr,
                "M98 codec diag: src=%lu emitted=%lu firstDiff=%lu "
                "src[d]=%c emitted[d]=%c err=%s\n",
                (unsigned long)src.length,
                (unsigned long)(emitted ? emitted.length : 0),
                (unsigned long)diff,
                diff < src.length ? a[diff] : '?',
                (emitted && diff < emitted.length) ? b[diff] : '?',
                [[err localizedDescription] UTF8String] ?: "none");
    }
    PASS([emitted isEqualToData:src],
         "M98 codec: BASE_PACK channel round-trips byte-exact");
}

// ── per-AU encryption + signatures (stage E) ──────────────────────

static NSData *m98Key(uint8_t seed)
{
    NSMutableData *k = [NSMutableData dataWithLength:32];
    uint8_t *b = k.mutableBytes;
    for (int i = 0; i < 32; i++) b[i] = (uint8_t)(seed + i);
    return k;
}

static NSString *m98WriteGraphFile(NSString *tag, NSData *src)
{
    NSError *err = nil;
    TTIOWrittenAssemblyGraph *g = [TTIOGfaReader graphFromData:src
                                                         error:&err];
    NSString *path = m98TmpPath(tag);
    m98Rm(path);
    if (![TTIOSpectralDataset writeMinimalToPath:path
                                           title:@"M98"
                             isaInvestigationId:@"ISA-M98"
                                         msRuns:@{}
                                     genomicRuns:nil
                                  assemblyGraphs:@{@"graph_0001": g}
                                 identifications:nil
                                 quantifications:nil
                               provenanceRecords:nil
                                           error:&err]) {
        return nil;
    }
    return path;
}

static void m98PerAUEncryption(void)
{
    // Workplan acceptance: a per-AU-encrypted graph decrypts in
    // place. Mechanism check: after encrypt the plaintext sequences
    // channel is GONE and the segments compound exists — a no-op
    // walker would still pass a bare round-trip.
    NSData *src = [m98SynthGfa() dataUsingEncoding:NSUTF8StringEncoding];
    NSString *path = m98WriteGraphFile(@"perau", src);
    PASS(path != nil, "M98 perau: fixture file written");
    if (!path) return;

    NSError *err = nil;
    PASS([TTIOPerAUFile encryptFilePath:path
                                    key:m98Key(0x42)
                         encryptHeaders:NO
                           providerName:nil
                                  error:&err],
         "M98 perau: encryptFilePath succeeds (%s)",
         [[err localizedDescription] UTF8String] ?: "");

    @autoreleasepool {
        TTIOHDF5File *f = [TTIOHDF5File openReadOnlyAtPath:path error:NULL];
        TTIOHDF5Group *seg =
            [[[[f.rootGroup openGroupNamed:@"study" error:NULL]
                openGroupNamed:@"assembly_graphs" error:NULL]
                    openGroupNamed:@"graph_0001" error:NULL]
                        openGroupNamed:@"segments" error:NULL];
        PASS(![seg hasChildNamed:@"sequences"]
                 && [seg hasChildNamed:@"sequences_segments"],
             "M98 perau: plaintext channel replaced by segments");
        [f close];
    }

    // The wrong key must not decrypt.
    NSError *wrongErr = nil;
    PASS(![TTIOPerAUFile decryptFilePathInPlace:path
                                            key:m98Key(0x99)
                                   providerName:nil
                                          error:&wrongErr],
         "M98 perau: wrong key rejected");

    PASS([TTIOPerAUFile decryptFilePathInPlace:path
                                           key:m98Key(0x42)
                                  providerName:nil
                                         error:&err],
         "M98 perau: decryptFilePathInPlace succeeds (%s)",
         [[err localizedDescription] UTF8String] ?: "");

    @autoreleasepool {
        TTIOHDF5File *f = [TTIOHDF5File openReadOnlyAtPath:path error:NULL];
        TTIOHDF5Group *seg =
            [[[[f.rootGroup openGroupNamed:@"study" error:NULL]
                openGroupNamed:@"assembly_graphs" error:NULL]
                    openGroupNamed:@"graph_0001" error:NULL]
                        openGroupNamed:@"segments" error:NULL];
        PASS([seg hasChildNamed:@"sequences"]
                 && ![seg hasChildNamed:@"sequences_segments"]
                 && ![seg hasAttributeNamed:@"sequences_algorithm"],
             "M98 perau: plaintext channel restored, segments removed");
        NSArray *features =
            [TTIOFeatureFlags featuresForRoot:f.rootGroup] ?: @[];
        PASS(![features containsObject:@"opt_per_au_encryption"],
             "M98 perau: opt_per_au_encryption stripped");
        [f close];
    }

    TTIOSpectralDataset *ds =
        [TTIOSpectralDataset readFromFilePath:path error:&err];
    PASS(ds != nil, "M98 perau: dataset reopens after decrypt");
    if (ds) {
        NSData *emitted =
            [ds.assemblyGraphs[@"graph_0001"] gfaDataWithError:&err];
        PASS([emitted isEqualToData:src],
             "M98 perau: decrypted graph re-emits byte-exact");
        [ds closeFile];
    }
    m98Rm(path);
}

static void m98Signatures(void)
{
    // v2: signatures cover segments/sequences the way they cover
    // genomic-run channels.
    NSData *src = [m98SynthGfa() dataUsingEncoding:NSUTF8StringEncoding];
    NSString *path = m98WriteGraphFile(@"sig", src);
    PASS(path != nil, "M98 sig: fixture file written");
    if (!path) return;

    NSError *err = nil;
    NSData *key = m98Key(0x10);
    NSDictionary *sigs = [TTIOSignatureManager
        signAssemblyGraph:@"graph_0001"
                   inFile:path
                  withKey:key
                    error:&err];
    PASS(sigs != nil && sigs[@"segments/sequences"] != nil,
         "M98 sig: segments/sequences signed (%s)",
         [[err localizedDescription] UTF8String] ?: "");
    PASS([sigs[@"segments/sequences"] hasPrefix:@"v2:"],
         "M98 sig: signature carries the v2: prefix");
    PASS([TTIOSignatureManager verifyAssemblyGraph:@"graph_0001"
                                            inFile:path
                                           withKey:key
                                             error:&err],
         "M98 sig: verification passes with the signing key");
    PASS(![TTIOSignatureManager verifyAssemblyGraph:@"graph_0001"
                                             inFile:path
                                            withKey:m98Key(0x77)
                                              error:NULL],
         "M98 sig: verification fails with the wrong key");
    m98Rm(path);
}

void testM98AssemblyGraph(void);
void testM98AssemblyGraph(void)
{
    m98ParseEmit();
    m98StorageRoundTrip();
    m98WriteMinimal();
    m98SequencesCodec();
    m98PerAUEncryption();
    m98Signatures();
}
